chain/client/src/sync/state.rs

```diff
@@ -359,39 +365,69 @@ impl StateSync {
     fn process_downloaded_parts(
         &mut self,
+        chain: &mut Chain,
         sync_hash: CryptoHash,
         shard_sync: &mut HashMap<u64, ShardSyncDownload>,
     ) {
-        for StateSyncGetPartResult { sync_hash: msg_sync_hash, shard_id, part_id, part_result } in
+        for StateSyncGetFileResult { sync_hash: msg_sync_hash, shard_id, part_id, result } in
             self.state_parts_mpsc_rx.try_iter()
         {
             if msg_sync_hash != sync_hash {
                 tracing::debug!(target: "sync",
                     ?shard_id,
-                    ?part_id,
                     ?sync_hash,
                     ?msg_sync_hash,
-                    "Received message for other sync.",
+                    "Received message for other epoch.",
                 );
                 continue;
             }
-            info!(target: "sync", ?part_result, ?part_id, ?shard_id, "downloaded a state part");
             if let Some(shard_sync_download) = shard_sync.get_mut(&shard_id) {
-                if shard_sync_download.status != ShardSyncStatus::StateDownloadParts {
-                    continue;
-                }
+                let download_result = match result {
+                    Err(err) => Err(err),
+                    // Store the header
+                    Ok(StateSyncFileDownloadResult::StateHeader { header_length, header }) => {
+                        info!(target: "sync", ?header_length, ?part_id, "processing state header");
+                        if shard_sync_download.status != ShardSyncStatus::StateDownloadHeader {
+                            continue;
+                        }
+                        if shard_sync_download
+                            .get_header_download_mut()
+                            .and_then(|d| Some(d.done))
+                            .unwrap_or(false)
+                        {
+                            continue;
+                        }
+                        chain
+                            .set_state_header(shard_id, sync_hash, header)
+                            .map_err(|err| format!("State sync set_state_header error: {err:?}"))
+                            .map(|_| header_length)
+                    }
+                    // Part was stored on the tx side.
+                    Ok(StateSyncFileDownloadResult::StatePart { part_length }) => {
+                        info!(target: "sync", ?part_length, ?part_id, ?shard_id, "processing state part");
+                        if shard_sync_download.status != ShardSyncStatus::StateDownloadParts {
+                            continue;
+                        }
+                        Ok(part_length)
+                    }
+                };
+                let (file_type, download_idx) = if let Some(part_id) = part_id {
+                    (StateFileType::part_str(), part_id.idx)
+                } else {
+                    (StateFileType::header_str(), 0)
+                };
```

Comment:
Here the info about the type is duplicated in the result enum and the part id optionality. Ideally it should only be one.

Reply:
fixed